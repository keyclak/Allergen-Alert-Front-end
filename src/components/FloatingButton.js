import React, { useEffect, useState } from 'react';
import { View, Text, Pressable, StyleSheet, ActivityIndicator, TextInput } from 'react-native';
import { Colors, StyleConstants, Styles } from '../style';

export default function FloatingButton({onPress, children}) {

    return (
        <Pressable style={[Styles.floatingButton, {bottom: StyleConstants.FloatingButtonSize / 2, right: StyleConstants.FloatingButtonSize / 2}]} onPress={onPress}>
            {children}
        </Pressable>
    )
}