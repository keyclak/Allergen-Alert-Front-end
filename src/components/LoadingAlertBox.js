import React, { useEffect, useRef, useState } from 'react';
import { View, Text, Pressable, StyleSheet, ActivityIndicator, TextInput, Animated } from 'react-native';
import { Colors, StyleConstants, Styles } from '../style';
import { MaterialIcons } from '@expo/vector-icons';

export default function LoadingAlertBox({text, icon, colors, loading, loadingText, error, errorIcon, errorColors}) {
    const animation = useRef(new Animated.Value(0)).current;

    function open() {
        Animated.timing(
            animation,
            {
                toValue: 1,
                duration: 300,
                useNativeDriver: true
            }
        ).start();
    }

    function startClose() {
        Animated.sequence([
            Animated.delay(3000),
            Animated.timing(
                animation,
                {
                    toValue: 0,
                    duration: 300,
                    useNativeDriver: true
                }
            )
        ]).start();
    }

    useEffect(() => {
        if(loading) {
            open();
        } else {
            startClose();
        }
    }, [loading])

    const fg = error ? errorColors[5] : colors[5];
    const bg = error ? errorColors[1] : colors[1];

    return (
        <Animated.View style={[
            Styles.alertBoxContainer,
            {
                borderColor: fg,
                backgroundColor: bg,
                transform: [
                    {
                        translateY: animation.interpolate({
                            inputRange: [0,1],
                            outputRange: [-100, 0]
                        })
                    }
                ]
            }]}>

            <View style={{width: 32}}>
            {
                loading
                    ? <ActivityIndicator color={colors[5]} size="large" />
                    : <MaterialIcons name={error ? errorIcon : icon} size={32} color={fg}/>
            }
            </View>

            <Text style={[Styles.alertBoxText, { color: fg }]}>{error || (loading ? loadingText : text)}</Text>
        </Animated.View>
    );
}