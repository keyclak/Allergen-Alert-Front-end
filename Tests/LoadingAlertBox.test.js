import renderer from 'react-test-renderer';
import React from 'react';
import MockDate from 'mockdate'
import FloatingButton from '../src/components/FloatingButton';
import LoadingAlertBox from '../src/components/LoadingAlertBox';

const FRAME_TIME = 10

const advanceOneFrame = () => {
  const now = Date.now();
  MockDate.set(new Date(now + FRAME_TIME));
  jest.advanceTimersByTime(FRAME_TIME);
}

const timeTravel = (msToAdvance = FRAME_TIME) => {
  const numberOfFramesToRun = msToAdvance / FRAME_TIME;
  let framesElapsed = 0;

  while (framesElapsed < numberOfFramesToRun) {
    advanceOneFrame();
    framesElapsed++;
  }
}
global.requestAnimationFrame = cb => {
  setTimeout(cb, FRAME_TIME);
}

export const setupTimeTravel = () => {
    MockDate.set(0);
    jest.useFakeTimers();
}

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
jest.mock('@expo/vector-icons', () => ({
    MaterialIcons: 'Icon'
}));

beforeEach(setupTimeTravel);

describe("FloatingButtonTests", () => {
    test("Renders correctly", async () => {
        const tree = renderer.create(
            <LoadingAlertBox text={jest.fn()} icon={jest.fn()} colors={jest.fn()} loading={jest.fn()} loadingText={jest.fn()} error={jest.fn()} errorIcon={jest.fn()} errorColors={jest.fn()} />
        ).toJSON(); 

        timeTravel(500);

        expect(tree).toMatchSnapshot();
    })
})