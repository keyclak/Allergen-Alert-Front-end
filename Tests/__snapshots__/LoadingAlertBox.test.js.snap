// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingButtonTests Renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": undefined,
      "borderColor": undefined,
      "borderRadius": 15,
      "borderWidth": 2,
      "elevation": 5,
      "flexDirection": "row",
      "height": 50,
      "paddingLeft": 5,
      "position": "absolute",
      "top": 30,
      "transform": Array [
        Object {
          "translateY": -100,
        },
      ],
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "width": 32,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
  <Text
    style={
      Array [
        Object {
          "fontSize": 16,
          "marginLeft": 5,
        },
        Object {
          "color": undefined,
        },
      ]
    }
  />
</View>
`;
